master/internal/trial.go

```diff
@@ -575,6 +579,38 @@ func (t *trial) handleAllocationExit(exit *task.AllocationExited) error {
 			WithError(exit.Err).
 			Errorf("trial encountered transient system error")
 	case exit.Err != nil && !sproto.IsTransientSystemError(exit.Err):
+		// First check against log_pattern_policies retries.
+		notRetries, err := logpattern.ShouldRetry(context.TODO(), t.taskID)
+		if err != nil {
+			return fmt.Errorf("getting if the trial should retry due to log_pattern_policies: %w", err)
+		}
+
+		if len(notRetries) > 0 {
+			var exitReasons []string
+			for _, r := range notRetries {
+				exitReasons = append(exitReasons,
+					fmt.Sprintf("(log '%q' matched regex %s)", r.TriggeringLog, r.Regex))
+			}
+			t.syslog.
+				WithError(exit.Err).
+				Errorf("trial failed and not retrying due to logs matching a don't retry policy" +
+					" check trial logs for more info")
+
+			reason := ""
+			for _, l := range append([]string{
+				"trial failed and matched logs to a don't retry policy",
+			}, exitReasons...) {
+				reason += l + "\n"
+				tasklogger.Insert(tasklogger.CreateLogFromMaster(t.taskID, model.LogLevelError, l))
+			}
```

Comment:
would be nice to have a helper API that gives the informational reason so when we support this for tasks it doesn't just get duped.